### PR TITLE
Pallet Multihop XCM-IBC to move DOT from Polkadot to Cosmos in one transaction.

### PR DESCRIPTION
**Pull request by RustNinja**
- [x] PR title is my best effort to provide summary of changes and has clear text to be part of release notes 
- [x] I marked PR by `misc` label if it should not be in release notes
- [x] I have linked Zenhub/Github or any other reference item if one exists
- [x] I was clear on what type of deployment required to release my changes (node, runtime, contract, indexer, on chain operation, frontend, infrastructure) if any in PR title or description
- [x] I waited and did best effort for `pr-workflow-check / draft-release-check` to finish with success(green check mark) with my changes
- [x] I have added at least one reviewer in reviewers list
- [x] I tagged(@) or used other form of notification of one person who I think can handle best review of this PR
- [x] I have proved that PR has no general regressions of relevant features and processes required to release into production

### CODE REVIEW

**Review by dzmitry-lahoda — APPROVED** [review]

File: code/parachain/frame/pallet-multihop-xcm-ibc/Cargo.toml [hunk 1]
```diff
@@ -20,8 +20,9 @@ package = "parity-scale-codec"
 version = "3.0.0"
 
 [dependencies]
-bech32-no_std = { workspace = true, default-features = false, features = ["strict"] }
+bech32-no_std = { version = "0.7.3", default-features = false, features = ["strict"] }
```
Comment (dzmitry-lahoda) [inline code/parachain/frame/pallet-multihop-xcm-ibc/Cargo.toml:23]: why not to change in workspace? we started to use it only for cosmos thing just now

File: code/parachain/runtime/picasso/src/lib.rs [hunk 2]
```diff
@@ -751,6 +751,18 @@ parameter_types! {
 	  pub MinReward: Balance = 10 * CurrencyId::unit::<Balance>();
 	  pub Stake: Balance = 10 * CurrencyId::unit::<Balance>();
 }
+parameter_types! {
+		pub const PalletXcmIbcInstanceId: u8 = 192; // PalletXcmIbc: pallet_xcm_ibc = 192,
```
Comment (dzmitry-lahoda) [inline code/parachain/runtime/picasso/src/lib.rs:755]: you can get id from pallet. see simulator tests.

Comment (dzmitry-lahoda) [reply code/parachain/runtime/picasso/src/lib.rs:755]: no need copy paste.

Comment (RustNinja) [reply code/parachain/runtime/picasso/src/lib.rs:757]: can you please point the place where i can use the pallet to pallet id?

File: code/parachain/runtime/picasso/src/xcmp.rs [hunk 3]
```diff
@@ -110,7 +194,18 @@ pub type XcmOriginToTransactDispatchOrigin = (
 	XcmPassthrough<RuntimeOrigin>,
 );
 
-pub type LocalAssetTransactor = MultiCurrencyAdapter<
+// pub type LocalAssetTransactor = MultiCurrencyAdapter<
```
Comment (dzmitry-lahoda) [inline code/parachain/runtime/picasso/src/xcmp.rs:197]: please remove commented out text

File: code/parachain/runtime/picasso/src/xcmp.rs [hunk 4]
```diff
@@ -85,8 +85,92 @@ pub type LocationToAccountId = (
...
+					),
+			} => id,
+			_ => return Err(location),
+		};
+
+		Ok(id.into())
+	}
+
+	fn reverse(who: AccountId) -> Result<MultiLocation, AccountId> {
+		// let m = MultiLocation { parents: 0, interior: X4(
```
Comment (dzmitry-lahoda) [inline code/parachain/runtime/picasso/src/xcmp.rs:164]: please comment why text is commented our or remove

File: code/parachain/runtime/picasso/src/xcmp.rs (hunk 3 above)
Comment (RustNinja) [reply code/parachain/runtime/picasso/src/xcmp.rs]: Done

File: code/parachain/runtime/picasso/src/lib.rs (hunk 2 above)
Comment (dzmitry-lahoda) [reply code/parachain/runtime/picasso/src/lib.rs:757]:
```
pub fn move_runtime_pallet<
	const OLD_NAME: &'static str,
	const MIGRATED_STORAGE_VERSION: u16,
	NewPallet: PalletInfoAccess + GetStorageVersion,
>() -> Weight {
	let new_pallet_name = <NewPallet as PalletInfoAccess>::name();
```

Comment (dzmitry-lahoda) [reply code/parachain/runtime/picasso/src/lib.rs:757]:
```
fn fun(){
	let _ = <Ibc as PalletInfo>::index().unwrap();
}
```

Comment (RustNinja) [reply code/parachain/runtime/picasso/src/lib.rs:757]: Thank you so much.

**Review by dzmitry-lahoda — DISMISSED** [review]

**Comment by dzmitry-lahoda** [comment]
@RustNinja please rebase to latest main

**Review by dzmitry-lahoda — DISMISSED** [review]

File: code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs [hunk 5]
```diff
@@ -106,12 +120,12 @@ pub mod pallet {
...
 	#[pallet::storage]
-	#[pallet::getter(fn chain_id_to_miltihop_route_path)]
+	#[pallet::getter(fn route_id_to_miltihop_path)]
 	pub type ChainIdToMiltihopRoutePath<T: Config> = StorageMap<
 		_,
 		Blake2_128Concat,
-		u128, //chain id
-		BoundedBTreeSet<(ChainInfo, BoundedVec<u8, T::ChainNameVecLimit>), T::MaxMultihopCount>, /* route to forward */
+		u128, /* chain id */
+		BoundedVec<(ChainInfo, BoundedVec<u8, T::ChainNameVecLimit>), T::MaxMultihopCount>, /* route to forward */
```
Comment (vmarkushin) [inline code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs:128]: I'd suggest creating separate types for that, like: 
```
type ChainId = u128;
type Route = BoundedVec<(ChainInfo, BoundedVec<u8, T::ChainNameVecLimit>), T::MaxMultihopCount>;
```

File: code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs [hunk 6]
```diff
@@ -420,43 +544,52 @@ pub mod pallet {
...
 			let result = pallet_ibc::Pallet::<T>::transfer(
 				signed_account_id.into(),
 				transfer_params,
 				asset_id.unwrap(), //TODO remove unwrap
 				(*amount).into(),
 				memo.clone(),
 			);
+
+			let memo_bytes =
+				memo.clone().map(|m| m.to_string().as_bytes().to_vec()).unwrap_or(Vec::new());
```
Comment (vmarkushin) [inline code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs:592]:
```suggestion
				memo.as_ref().map(|m| m.to_string().into_bytes()).unwrap_or_default();
```

File: code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs [hunk 7]
```diff
@@ -465,6 +598,8 @@ pub mod pallet {
 						amount: *amount,
 						asset_id: asset_id.unwrap(),
 						memo,
+						bytes: memo_bytes.clone(),
```
Comment (vmarkushin) [inline code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs:601]: Do we need to clone here?

**Review by dzmitry-lahoda — APPROVED** [review]

File: code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs (hunk 5 above)
Comment (dzmitry-lahoda) [reply code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs:129]: type ChainId = u32;

**Review by dzmitry-lahoda — DISMISSED** [review]

**Comment by RustNinja** [comment]
> If it is not final PR, than do nothing, but if it is, please fix back linter - that is part of Peer Review Audit tbh. XC code i super hard to debug, tbh I would makke this event stricter (there are some codebases in our repo)
> 
> ```
> // #![cfg_attr(
> // 	not(test),
> // 	deny(
> // 		clippy::disallowed_methods,
> // 		clippy::disallowed_types,
> // 		clippy::indexing_slicing,
> // 		clippy::todo,
> // 		clippy::unwrap_used,
> // 		clippy::panic
> // 	)
> // )] // allow in tests
> // #![deny(clippy::unseparated_literal_suffix, unused_imports, dead_code)]
> ```

thanks. good point. will create next PR with clippy.
need to merge to update rococo to allow FE team to work on multihop functional.

**Review by blasrodri — APPROVED** [review]

**Review by blasrodri — APPROVED** [review]

**Review by dzmitry-lahoda — APPROVED** [review]

**Review by kkast — APPROVED** [review]

File: code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs (hunk 7 above)
Comment (RustNinja) [reply code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs:596]: will remove in next pr. thanks.

File: code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs (hunk 6 above)
Comment (RustNinja) [reply code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs:587]: thanks. will commit in next pr.

File: code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs (hunk 5 above)
Comment (RustNinja) [reply code/parachain/frame/pallet-multihop-xcm-ibc/src/lib.rs:127]: Thanks. I will create pr with this changes.

**Pull request merged**